migrations/20230613116515_courses.sql

```diff
@@ -0,0 +1,103 @@
+    title TEXT NOT NULL,
+    description TEXT NOT NULL,
+    abbreviated_description TEXT NOT NULL,
+    requirements_description TEXT NOT NULL,
+    prerequisites_id INTEGER NULL,
+    CONSTRAINT uk_courses_external_id UNIQUE (external_id),
+    CONSTRAINT fk_courses_to_prerequisites FOREIGN KEY (prerequisites_id)
+    REFERENCES prerequisites (id),
+    CONSTRAINT chk_courses_external_id_not_empty CHECK (external_id <> ''),
+    CONSTRAINT chk_courses_subject_code_not_empty CHECK (subject_code <> ''),
+    CONSTRAINT chk_courses_academic_level CHECK (academic_level IN ('undergraduate', 'graduate')),
+    CONSTRAINT chk_courses_title_not_empty CHECK (title <> '')
+);
+
+-- Courses Table Indices
+CREATE INDEX CONCURRENTLY IF NOT EXISTS idx_courses_subject_code_catalog_number
+    ON courses (subject_code, catalog_number);
+
+CREATE INDEX CONCURRENTLY IF NOT EXISTS idx_courses_title
+    ON courses (title);
+
+CREATE INDEX CONCURRENTLY IF NOT EXISTS idx_courses_academic_level
+    ON courses (academic_level);
+
+-- Course Offerings Table
+CREATE TABLE IF NOT EXISTS course_offerings (
+    id SERIAL PRIMARY KEY,
+    course_id INTEGER NOT NULL,
+    year SMALLINT NOT NULL,
+    term VARCHAR(6) NOT NULL,
+    start_date TIMESTAMP WITH TIME ZONE NOT NULL,
+    end_date TIMESTAMP WITH TIME ZONE NOT NULL,
+    CONSTRAINT uk_course_offerings_unique_data UNIQUE (course_id, year, term),
+    CONSTRAINT fk_course_offerings_to_courses FOREIGN KEY (course_id)
+        REFERENCES courses (id)
+        ON DELETE CASCADE,
+    CONSTRAINT chk_course_offerings_term CHECK (term IN ('winter', 'fall', 'spring')),
+    CONSTRAINT chk_course_offerings_end_date CHECK (end_date > start_date)
+    );
+
+-- Course Offerings Table Indices
+CREATE INDEX CONCURRENTLY IF NOT EXISTS idx_course_offerings_course_id
+    ON course_offerings (course_id);
+
+CREATE INDEX CONCURRENTLY IF NOT EXISTS idx_course_offerings_year_term
+    ON course_offerings (year, term);
+
+CREATE TABLE IF NOT EXISTS prerequisites (
+    id SERIAL PRIMARY KEY,
+    logical_operator VARCHAR(3) NOT NULL,
+    CONSTRAINT chk_prerequisites_logical_operator CHECK (logical_operator IN ('AND', 'OR'))
+);
+
+CREATE TABLE IF NOT EXISTS prerequisite_courses (
+    prerequisite_id INTEGER NOT NULL,
+    sub_prerequisite_id INTEGER NULL,
+    course_id INTEGER NULL,
+    minimum_course_grade NUMERIC(5, 2) NULL,
+    CONSTRAINT chk_prerequisite_courses_minimum_course_grade CHECK (
+        minimum_course_grade >= 0.0 AND minimum_course_grade <= 100.0
+    ),
+    CONSTRAINT fk_prerequisite_courses_to_prerequisites FOREIGN KEY (prerequisite_id)
+        REFERENCES prerequisites (id)
+        ON DELETE CASCADE,
+    CONSTRAINT fk_sub_prerequisite_courses_to_prerequisites FOREIGN KEY (sub_prerequisite_id)
+        REFERENCES prerequisites (id)
+        ON DELETE SET NULL,
+    CONSTRAINT fk_prerequisite_courses_to_courses FOREIGN KEY (course_id)
+        REFERENCES courses (id)
+        ON DELETE SET NULL
```

Comment:
Prerequisite course combination can require multple other sub combinations and/or a single course (leaf node).

Example: `(CS135 AND CS136) OR CS246`